cli/flox-rust-sdk/src/models/environment/core_environment.rs

```diff
@@ -507,12 +589,80 @@ impl CoreEnvironment<ReadOnly> {
         )
         .map_err(CoreEnvironmentError::ParseUpgradeOutput)?;
 
-        let store_path = self.transact_with_lockfile_contents(json.lockfile.to_string(), flox)?;
+        Ok((json.lockfile, json.result.0))
+    }
 
-        Ok(UpgradeResult {
-            packages: json.result.0,
-            store_path: Some(store_path),
-        })
+    /// Upgrade the given groups or install ids in the environment using the catalog client.
+    /// The environment is upgraded by locking the existing manifest
+    /// using [LockedManifestCatalog::lock_manifest] with the existing lockfile as a seed,
+    /// where the upgraded packages have been filtered out causing them to be re-resolved.
+    fn upgrade_with_catalog_client(
+        &mut self,
+        client: &impl ClientTrait,
+        groups_or_iids: &[String],
+        manifest: &TypedManifestCatalog,
+    ) -> Result<
+        (
+            LockedManifestCatalog,
+            Vec<(LockedPackageCatalog, LockedPackageCatalog)>,
+        ),
+        CoreEnvironmentError,
+    > {
+        let existing_lockfile = 'lockfile: {
+            let Ok(lockfile_path) = CanonicalPath::new(self.lockfile_path()) else {
+                break 'lockfile None;
+            };
+            let lockfile = LockedManifest::read_from_file(&lockfile_path)
+                .map_err(CoreEnvironmentError::LockedManifest)?;
+            match lockfile {
+                LockedManifest::Catalog(lockfile) => Some(lockfile),
+                _ => {
+                    warn!(
+                        "Found version 1 manifest, but lockfile doesn't match: Ignoring lockfile."
+                    );
+                    None
+                },
+            }
+        };
+
+        let previous_packages = existing_lockfile
+            .as_ref()
+            .map(|lockfile| lockfile.packages.clone())
+            .unwrap_or_default();
+
+        // Create a seed lockfile by "unlocking" (i.e. removing the locked entries of)
+        // all packages matching the given groups or iids.
+        // If no groups or iids are provided, all packages are unlocked.
+        let seed_lockfile = if groups_or_iids.is_empty() {
+            debug!("no groups or iids provided, not unlocking all packages");
```

Comment:
I think that means not calling the `unlock` function? But they're being unlocked, right?
```suggestion
            debug!("no groups or iids provided, unlocking all packages");
```